### PR TITLE
Edit comprehend entities and keep analysis pane open during and after transcription

### PR DESCRIPTION
**Pull request by ShivaniMehendarge**
*Issue #, if available:*

*Description of changes:*
1. Allow adding and deleting of comprehend entities
2. Keep analysis pane open during and after transcription
3. Allow editing transcription after the audio/recording is complete

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: app/src/components/AnalysisPane.js [hunk 1]
```diff
@@ -1,147 +1,194 @@
...
-        'RxNorm Concepts',
-        <ConceptTable rows={result.RxNormConcepts} />
-      ]);
-    }
-
     return a;
-  }, [ result ]);
+  }, [result]);
+
+  if (!result.ICD10CMConcepts && !result.RxNormConcepts && !result.Attributes) {
```
Comment (alexchirayath) [inline app/src/components/AnalysisPane.js]: Can both these if conditions be merged together or refactored to avoid duplication?
```
If (!ICD && !RXnorm){
if (result.Attr)
else {
}

}
```

File: app/package.json [hunk 2]
```diff
@@ -107,6 +112,9 @@
     "websocket-extensions": "^0.1.4",
     "yargs-parser": "^18.1.3"
   },
+  "resolutions": {
```
Comment (kazbaig) [inline app/package.json]: Why was this added?

File: app/src/components/AnalysisPane.js [hunk 3]
```diff
@@ -1,147 +1,194 @@
...
-        {attrs.map(([ key, value ]) => (
-          <>
-            <dt>{key}</dt>
-            <dd>{value}</dd>
-          </>
+    <Flex width='100%' alignItems='center'>
+      <Select
+        mr={2}
+        border='1px solid'
+        borderColor={concepts[0].Score < 0.5 ? 'red' : 'grey'}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Can extract `concepts[0].Score < 0.5 ? 'red' : 'grey'` to a variable for reuse below

File: app/src/useComprehension.js [hunk 4]
```diff
@@ -19,42 +19,42 @@ const resultMap = new WeakMap();
 // React hook to take an array of transcript chunks and return a corresponding array of Comprehend results, one for each
 export default function useComprehension(transcriptChunks, clientParams) {
   // Dummy state variable so we can force a refresh when we receive an update
```
Comment (kazbaig) [inline app/src/useComprehension.js]: This comment should be removed as it does not apply to the new state

File: app/src/components/AnalysisPane.js [hunk 5]
```diff
@@ -1,147 +1,194 @@
...
+  }, [result]);
+
+  if (!result.ICD10CMConcepts && !result.RxNormConcepts && !result.Attributes) {
+    return (
+      <Flex width='100%' alignItems='center'>
+        <Flex
+          flex='1'
+          mr={2}
+          height='2.5rem'
+          border={result.Score && result.Score < 0.5 ? '1px solid red' : '1px solid grey'}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Edge case - it will be grey instead of red when `result.Score` is `0` because `0` is falsy. Is that okay with us? Will we ever receive a score of `0`?

File: app/src/components/AnalysisPane.js [hunk 6]
```diff
@@ -1,147 +1,194 @@
...
+            mr={2}
+            mb={4}
+            border='1px solid'
+            borderColor='grey'
+            borderRadius='0'
+            bg='white'
+            placeholder={'Add ' + displayNames[category]}
+            value={inputValue}
+            onChange={handleInputChange}
+          ></Input>
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Components/elements not receiving a `children` prop should be self-closing in JSX.

File: app/src/components/AnalysisPane.js [hunk 7]
```diff
@@ -1,147 +1,194 @@
...
+      <VStack spacing={2}>
+        <Flex width='100%' alignItems='center'>
+          <Input
+            mr={2}
+            mb={4}
+            border='1px solid'
+            borderColor='grey'
+            borderRadius='0'
+            bg='white'
+            placeholder={'Add ' + displayNames[category]}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Template syntax is idiomatically preferred over string concatenation

File: app/src/components/AnalysisPane.js [hunk 8]
```diff
@@ -1,147 +1,194 @@
...
+    <Box className={s.resultTable} mb={4}>
+      <Box as='h1' mb={4} textAlign='left' fontWeight='bold' fontSize='1.2rem'>
+        {displayNames[category]}
+      </Box>
+
+      <VStack spacing={2}>
+        <Flex width='100%' alignItems='center'>
+          <Input
+            mr={2}
+            mb={4}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: This margin is causing misalignment with the button. It should go on their parent.

Screen Shot 2021-03-08 at 2 21 40 PM

File: app/src/components/AnalysisPane.module.css [hunk 9]
```diff
@@ -54,7 +50,8 @@
 }
 
 .resultTable {
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.module.css]: Since the user of this class is now a chakra component, let's use style props to remove this css declaration rather than mixing style props and css.

File: app/src/components/AnalysisPane.js [hunk 10]
```diff
@@ -1,147 +1,194 @@
...
+    </Flex>
+  );
 }
 
+function ResultTable({ results, category, onToggleItem, excludedItems, onResultDelete, onResultAdd }) {
+  const filteredResults = useMemo(() => results.filter((r) => r.Category === category), [results, category]);
+  const [inputValue, setInputValue] = React.useState('');
+  const handleInputChange = (event) => setInputValue(event.target.value);
+
+  const addIcon = (
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js:150]: The hover state is really difficult to see against the background (bad for accessibility). Let's work with design to improve it.

File: app/src/components/AnalysisPane.js [hunk 11]
```diff
@@ -1,147 +1,194 @@
...
+          mr={2}
+          height='2.5rem'
+          border={result.Score && result.Score < 0.5 ? '1px solid red' : '1px solid grey'}
+          bg='white'
+          px={4}
+          alignItems='center'
+        >
+          {attrs.map(([key, value]) => (
+            <React.Fragment key={key}>
+              {result.Text} {value ? '|' : null} {value}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: `{ booleanyValue && <Component /> }` is preferred for conditional rendering

File: app/src/components/AnalysisPane.js (hunk 1 above)
Comment (kazbaig) [reply app/src/components/AnalysisPane.js]: To add to this, moving the expression in the `if` statements to variables would help understand why the condition results in the returned value

File: app/src/components/AnalysisPane.js [hunk 12]
```diff
@@ -1,147 +1,194 @@
...
+            </React.Fragment>
+          ))}
+        </Flex>
+
+        {closeIcon}
+      </Flex>
+    );
+  }
 
+  const concepts = [...(result.ICD10CMConcepts ? result.ICD10CMConcepts : result.RxNormConcepts)];
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: IMO this is more concise
```js
const concepts = [...(result.ICD10CMConcepts ? result.ICD10CMConcepts : result.RxNormConcepts)].sort(
    (concept1, concept2) => concept2.Score - concept1.Score,
  );
```

File: app/src/components/AnalysisPane.js [hunk 13]
```diff
@@ -1,147 +1,194 @@
...
+        mr={2}
+        border='1px solid'
+        borderColor={concepts[0].Score < 0.5 ? 'red' : 'grey'}
+        borderRadius='0'
+        bg='white'
+        _hover={{ borderColor: concepts[0].Score < 0.5 ? 'red' : 'grey', boxShadow: 'none' }}
+      >
+        {concepts.map((concept) => (
+          <option key={concept.Code} value={concept.Code}>
+            {result.Text} {concept.Code ? ' | ' : null} {concept.Code} {concept.Description ? ' | ' : null}{' '}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Same comment about conditional rendering

File: app/src/components/AnalysisPane.js [hunk 14]
```diff
@@ -1,147 +1,194 @@
...
+    <Flex width='100%' alignItems='center'>
+      <Select
+        mr={2}
+        border='1px solid'
+        borderColor={concepts[0].Score < 0.5 ? 'red' : 'grey'}
+        borderRadius='0'
+        bg='white'
+        _hover={{ borderColor: concepts[0].Score < 0.5 ? 'red' : 'grey', boxShadow: 'none' }}
+      >
+        {concepts.map((concept) => (
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Destructuring the `(concept) =>` to `({ Code, Text, Description }) => ` should make this a bit cleaner

File: app/src/components/AnalysisPane.js [hunk 15]
```diff
@@ -1,147 +1,194 @@
...
+          <option key={concept.Code} value={concept.Code}>
+            {result.Text} {concept.Code ? ' | ' : null} {concept.Code} {concept.Description ? ' | ' : null}{' '}
+            {concept.Description} &nbsp;&nbsp; {(concept.Score * 100).toPrecision(4)}%
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Is `concept.Score` potentially undefined? `undefined * 100` will be `NaN`.

File: app/src/components/AnalysisPane.js [hunk 16]
```diff
@@ -1,147 +1,194 @@
...
+function ResultTable({ results, category, onToggleItem, excludedItems, onResultDelete, onResultAdd }) {
+  const filteredResults = useMemo(() => results.filter((r) => r.Category === category), [results, category]);
+  const [inputValue, setInputValue] = React.useState('');
+  const handleInputChange = (event) => setInputValue(event.target.value);
+
+  const addIcon = (
+    <IconButton
+      aria-label='Add'
+      icon={<AddIcon />}
+      onClick={() => [inputValue.trim() !== '' ? onResultAdd(inputValue, category) : null, setInputValue('')]}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Why is this an array? Looks like it should be a function that calls 2 functions.

Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: We should call `onResultAdd` with the trimmed value as a better UX. Also, the input should only be cleared when the result was added. Here, I could add a bunch of whitespace and it would cleared even if we decided not to add.

File: app/src/components/ExportPane.js [hunk 17]
```diff
@@ -3,73 +3,56 @@ import React, { useMemo } from 'react';
...
 import cs from 'clsx';
 
-
 import displayNames from '../displayNames';
 
-function ResultRow({
-  result
-}) {
-
+function ResultRow({ result }) {
```
Comment (kazbaig) [inline app/src/components/ExportPane.js]: Are there any changes in this file besides the addition of the key or is it just all formatting?

File: app/src/home.js [hunk 18]
```diff
@@ -567,22 +577,43 @@ export default function Home() {
...
+        const a = prevResult.findIndex((result) => result.id === r.id);
+
+        if (a === -1) return prevResult;
+
+        return [...prevResult.slice(0, a), ...prevResult.slice(a + 1)];
+      }),
+    );
+  };
+
+  const onComprehendResultAddition = (val, category) => {
```
Comment (kazbaig) [inline app/src/home.js:637]: I would expect the addition to show as the first item in the list, right below the input element. Let's see how others feel.

File: app/src/components/AnalysisPane.js [hunk 19]
```diff
@@ -1,147 +1,194 @@
...
-
+  return (
+    <Box className={s.resultTable} mb={4}>
+      <Box as='h1' mb={4} textAlign='left' fontWeight='bold' fontSize='1.2rem'>
+        {displayNames[category]}
+      </Box>
+
+      <VStack spacing={2}>
+        <Flex width='100%' alignItems='center'>
+          <Input
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: [Input elements should always have labels, even if they're hidden visually.](https://www.w3.org/WAI/tutorials/forms/labels/). You can use the [`VisuallyHidden`](https://chakra-ui.com/docs/disclosure/visually-hidden) component to achieve this.

**Review by kazbaig — CHANGES_REQUESTED** [review]

File: app/src/useComprehension.js (hunk 4 above)
Comment (ShivaniMehendarge) [reply app/src/useComprehension.js]: Will do.

File: app/src/components/AnalysisPane.js (hunk 3 above)
Comment (alexchirayath) [reply app/src/components/AnalysisPane.js]: To add on to that,  you could think of moving the confidence score threshold to a constant value. We may want to enable the user in the future to enter their custom defined threshold

File: app/src/components/AnalysisPane.js (hunk 5 above)
Comment (ShivaniMehendarge) [reply app/src/components/AnalysisPane.js]: I haven't encountered any instance where the score is 0. the model should not detect the entity if it has 0 confidence in it ideally right?

File: app/src/components/AnalysisPane.js (hunk 8 above)
Comment (ShivaniMehendarge) [reply app/src/components/AnalysisPane.js]: Got it. Thanks!

File: app/src/components/AnalysisPane.module.css (hunk 9 above)
Comment (ShivaniMehendarge) [reply app/src/components/AnalysisPane.module.css]: Will do.

File: app/src/components/AnalysisPane.js (hunk 5 above)
Comment (kazbaig) [reply app/src/components/AnalysisPane.js]: Probably. Just want to make sure we've considered it first.

File: app/src/components/AnalysisPane.js (hunk 16 above)
Comment (ShivaniMehendarge) [reply app/src/components/AnalysisPane.js]: Good point.

File: app/src/components/AnalysisPane.js [hunk 20]
```diff
@@ -1,147 +1,194 @@
...
-      <button className={s.removeButton} onClick={onHide} aria-label="Exclude this item from the export" />
-      <h4>{result.Text}</h4>
-
-      <dl>
-        {attrs.map(([ key, value ]) => (
-          <>
-            <dt>{key}</dt>
-            <dd>{value}</dd>
-          </>
+    <Flex width='100%' alignItems='center'>
```
Comment (alexchirayath) [inline app/src/components/AnalysisPane.js:101]: If we are allowing them to choose the dropdown to select another code/item for that specific entity, is there a way to record that selection? That selection can be then used to generate the summary page. For example, can we move that selection to the top of the list for ICD10/RX norm?

Comment (kazbaig) [reply app/src/components/AnalysisPane.js:101]: We said that we would remember the chosen concept once we decide that we're going to print it in the summary (we currently don't). Now that we have a PR open that includes that change, let's add the state piece here and use it in the summary gen PR.

File: app/package.json (hunk 2 above)
Comment (ShivaniMehendarge) [reply app/package.json]: It was not supposed to be added. Will remove.

File: app/src/components/AnalysisPane.js (hunk 19 above)
Comment (ShivaniMehendarge) [reply app/src/components/AnalysisPane.js]: Added.

File: app/src/home.js (hunk 18 above)
Comment (ShivaniMehendarge) [reply app/src/home.js:637]: Updated

File: app/src/components/ExportPane.js (hunk 17 above)
Comment (ShivaniMehendarge) [reply app/src/components/ExportPane.js]: Nope. Nothing besides those changes

File: app/src/components/AnalysisPane.js [hunk 21]
```diff
@@ -1,147 +1,233 @@
...
+
+function ResultRow({ result, stage, onToggleItem, excludedItems, onDeleteClick }) {
+  const closeIcon = (
+    <IconButton
+      aria-label='Delete'
+      icon={<DeleteIcon />}
+      onClick={onDeleteClick}
+      size='xs'
+      isRound
+      isDisabled={stage === STAGE_TRANSCRIBING && true}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: `&& true` is extra here

File: app/src/components/AnalysisPane.js [hunk 22]
```diff
@@ -1,147 +1,233 @@
...
-  onToggleItem,
-  excludedItems
-}) {
-
-  const onHide = useCallback(() => {
-    onToggleItem(result.id)
-  }, [ onToggleItem, result.id ]);
-
-  const isExcluded = useMemo(() => excludedItems.includes(result.id), [ excludedItems, result.id ]);
+const confidenceThreshold = 0.5;
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: Let's move this to the `consts` file or a new similarly named file and use CONSTANT_CASE since this is technically a global constant

File: app/src/components/AnalysisPane.js [hunk 23]
```diff
@@ -1,147 +1,233 @@
...
+  };
+
+  const addIcon = (
+    <IconButton
+      aria-label='Add'
+      icon={<AddIcon />}
+      onClick={handleSubmit}
+      size='xs'
+      isRound
+      isDisabled={stage === STAGE_TRANSCRIBING && true}
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: `&& true` is extra here

File: app/src/components/AnalysisPane.js [hunk 24]
```diff
@@ -1,147 +1,233 @@
...
+        fontSize='1.2rem'
+        className={highlightClasses[category]}
+      >
+        {displayNames[category]}
+      </Box>
+
+      <VStack spacing={2}>
+        <FormControl>
+          <Flex width='100%' mb={4} alignItems='center'>
+            <VisuallyHidden>`Add ${displayNames[category]}`</VisuallyHidden>
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: This should be `Add {displayNames[category]}`

Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: The rendered tag should be a `label` here, not the default `span` and should associate with the input via `htmlFor`

File: app/src/home.js [hunk 25]
```diff
@@ -565,24 +578,57 @@ export default function Home() {
...
     stage = STAGE_TRANSCRIBED;
   } else if (!showExport) {
     stage = STAGE_SUMMARIZE;
   } else {
     stage = STAGE_EXPORT;
   }
 
+  const onComprehendResultDelete = (r) => {
+    setComprehendResults((prevResults) =>
+      prevResults.map((prevResult) => {
```
Comment (kazbaig) [inline app/src/home.js]: One problem with always returning the result of `Array.map` in a setState function is that it will always trigger a rerender, even if we didn't actually change any inner values. Let's see if there's something we actually need to change before mapping and return the previous value if not for these setState calls here.

File: app/src/home.js [hunk 26]
```diff
@@ -565,24 +578,57 @@ export default function Home() {
...
+
+        if (a === -1) return prevEntity;
+
+        return [...prevEntity.slice(0, a), ...prevEntity.slice(a + 1)];
+      }),
+    );
+  };
+
+  const onComprehendResultAddition = (val, category) => {
+    setComprehendCustomEntities((prevEntities) => {
```
Comment (kazbaig) [inline app/src/home.js:638]: IMO we should have a field to identify custom entities as such as well - something like `isCustomEntity: true` - when initializing a custom entity so that we can avoid iterating through the comp med entities when we really want to just touch the custom entities and vice versa. This will also ensure that we don't accidentally delete the wrong entity based on id collision.

**Review by kazbaig — CHANGES_REQUESTED** [review]
Any thoughts on whether we should make the spacing even between the `input` and the entities?

Screen Shot 2021-03-09 at 1 15 05 PM

**Comment by ShivaniMehendarge** [comment]
> Any thoughts on whether we should make the spacing even between the `input` and the entities?
> 
> Screen Shot 2021-03-09 at 1 15 05 PM

Brenda had suggested that we have extra space between the input element and the entities. I don't mind having it either way.

**Comment by kazbaig** [comment]
> > Any thoughts on whether we should make the spacing even between the `input` and the entities?
> > Screen Shot 2021-03-09 at 1 15 05 PM
> 
> Brenda had suggested that we have extra space between the input element and the entities. I don't mind having it either way.

If this was intentional, I'm good!

File: app/src/components/AnalysisPane.js (hunk 22 above)
Comment (ShivaniMehendarge) [reply app/src/components/AnalysisPane.js]: Good point. Made the changes, Will push in a new commit

File: app/src/home.js (hunk 25 above)
Comment (ShivaniMehendarge) [reply app/src/home.js]: As per our conversation offline, we decided to leave this as is for now.

File: app/src/home.js (hunk 26 above)
Comment (ShivaniMehendarge) [reply app/src/home.js:638]: Good point. Added that in.

**Review by alexchirayath — APPROVED** [review]
LGTM

**Review by kazbaig — APPROVED** [review]

File: app/src/components/AnalysisPane.js [hunk 27]
```diff
@@ -1,126 +1,247 @@
...
-        ) : null}
-      </HeightSlider>
-    </div>
+      </Box>
+
+      <VStack spacing={2}>
+        <FormControl as='form' onSubmit={handleSubmit}>
+          <Flex width='100%' mb={4} alignItems='center'>
+            <VisuallyHidden as='label' htmlFor={addEntityInputId}>
+              Add displayNames[category]
```
Comment (kazbaig) [inline app/src/components/AnalysisPane.js]: This should be an expression, not a string

File: app/src/home.js [hunk 28]
```diff
@@ -577,27 +601,81 @@ export default function Home() {
...
+  } else if (!showSOAPReview) {
     stage = STAGE_TRANSCRIBED;
   } else if (!showExport) {
-    stage = STAGE_SUMMARIZE;
+    stage = STAGE_SOAP_REVIEW;
   } else {
     stage = STAGE_EXPORT;
   }
 
+  const onComprehendResultDelete = (r) => {
```
Comment (kazbaig) [inline app/src/home.js:615]: Let's flip the boolean for the ternary and remove console logs

File: app/src/home.js [hunk 29]
```diff
@@ -171,28 +173,55 @@ export default function Home() {
...
-        const itemContent = isPronunciation || isPunctuation ? item.Content : '';
-        const spaceAtEnd = isPronunciation ? ' ' : '';
-        const text = `${prevText}${speakerLabel}${itemContent}${spaceAtEnd}`;
-        const addSpeakerLabel = isSpeakerChange || (shouldAddSpeakerLabel ? false : prevAddSpeakerLabel);
-
-        return [text, addSpeakerLabel];
-      },
-      ['', true],
-    );
+    let addSpeakerLabel = true;
```
Comment (alexchirayath) [inline app/src/home.js]: This entire codeblock needs to the one that is on master. Please revert thie addTranscriptChunk to what is on master

File: app/src/consts.js [hunk 30]
```diff
@@ -4,6 +4,10 @@ export const STAGE_HOME = 0;
 export const STAGE_TRANSCRIBING = 1;
 export const STAGE_TRANSCRIBED = 2;
 export const STAGE_SUMMARIZE = 3;
```
Comment (alexchirayath) [inline app/src/consts.js]: I dont think we have a STAGE_SUMMARIZE any more.

**Review by alexchirayath — CHANGES_REQUESTED** [review]
Comments left inline

File: app/src/consts.js (hunk 30 above)
Comment (ShivaniMehendarge) [reply app/src/consts.js]: Should we rename the stages below it too then?

File: app/src/home.js (hunk 29 above)
Comment (ShivaniMehendarge) [reply app/src/home.js]: Reverted. Will push a new change

File: app/src/home.js (hunk 28 above)
Comment (ShivaniMehendarge) [reply app/src/home.js:615]: Done.

File: app/src/components/AnalysisPane.js (hunk 27 above)
Comment (ShivaniMehendarge) [reply app/src/components/AnalysisPane.js]: Good catch. Fixed it.

**Review by alexchirayath — APPROVED** [review]
